src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -325,6 +330,131 @@ EnsureCitusTableCanBeCreated(Oid relationOid)
 }
 
 
+/*
+ * undistribute_table gets a distributed table name and
+ * udistributes the table.
+ */
+Datum
+undistribute_table(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid relationId = PG_GETARG_OID(0);
+
+	if (!IsCitusTable(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute relation."),
+						errdetail("The relation is not distributed.")));
+	}
+
+	EnsureTableOwner(relationId);
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
+
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute relation"),
+						errdetail("No such distributed relation exists. "
+								  "Might have already been undistributed.")));
+	}
+
+	char *relationName = get_rel_name(relationId);
+
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+
+
+	if (TableReferencing(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because it has a foreign key.")));
+	}
+
+	if (TableReferenced(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because a foreign key references to it.")));
+	}
+
+	if (PartitionedTable(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute partitioned tables.")));
+	}
```

Comment:
It would be really nice if we could re-create the partitioning hierarchy here.

Note: The reason I'm being picky about this PR is that we want to introduce infrastructure to make arbitrary conversions between table types (e.g. change distribution column), which all comes down to creating a new table, insert..select, drop old table, rename, create indexes. I figure that if we spend a bit more time making this implementation really solid, the rest will be a breeze.